src/Simulation/Simulators/SparseSimulator/Native/SparseSimulator.h

```diff
@@ -0,0 +1,851 @@
+		// Commuting with Rx creates a phase, but on the wrong side
+		// So we execute any Rx immediately
+		if (_queue_Rx[index]){
+			_execute_queued_ops(index, OP::Rx);
+		}
+		_queue_H[index] = !_queue_H[index];
+		_set_qubit_to_nonzero(index);
+	}
+
+	void MCH(std::vector<logical_qubit_id> const& controls, logical_qubit_id target) {
+		// No commutation on controls
+		_execute_if(controls);
+		// No Ry or Rx commutation on target
+		if (_queue_Ry[target] || _queue_Rx[target]){
+			_execute_queued_ops(target, OP::Ry);
+		} 
+		// Commutes through H gates on the target, so it does not check
+		_execute_phase_and_permute();
+		_quantum_state->MCH(controls, target);
+		_set_qubit_to_nonzero(target);
+	}
+
+
+	
+
+	void SWAP(logical_qubit_id index_1, logical_qubit_id index_2){
+		// This is necessary for the "shift" to make sense
+		if (index_1 > index_2){
+			std::swap(index_2, index_1);
+		} 
+		// Everything commutes nicely with a swap
+		_queue_Ry.swap(_queue_Ry[index_1],  _queue_Ry[index_2]);
+		std::swap(_angles_Ry[index_1], _angles_Ry[index_2]);
+		_queue_Rx.swap(_queue_Rx[index_1],  _queue_Rx[index_2]);
+		std::swap(_angles_Rx[index_1], _angles_Rx[index_2]);
+		_queue_H.swap(_queue_H[index_1],  _queue_H[index_2]);
+		_occupied_qubits.swap(_occupied_qubits[index_1], _occupied_qubits[index_2]);
+		logical_qubit_id shift = index_2 - index_1;
+		_queued_operations.push_back(operation(OP::SWAP, index_1, shift, index_2));
+	}
+
+	void CSWAP(std::vector<logical_qubit_id> const& controls, logical_qubit_id index_1, logical_qubit_id index_2){
+		if (index_1 > index_2){
+			std::swap(index_2, index_1);
+		}
+		// Nothing commutes nicely with a controlled swap
+		_execute_if(controls);
+		_execute_if(index_1);
+		_execute_if(index_2);	
+
+		logical_qubit_id shift = index_2 - index_1;
+		_queued_operations.push_back(operation(OP::MCSWAP, index_1, shift, controls, index_2));
+		// If either qubit is occupied, then set them both to occupied
+		if(_occupied_qubits[index_1] || _occupied_qubits[index_2]){
+			_set_qubit_to_nonzero(index_1);
+			_set_qubit_to_nonzero(index_2);
+		}
+	}
+
+	bool M(logical_qubit_id target) {
+		// Do nothing if the qubit is known to be 0
+		if (!_occupied_qubits[target]){ 
+			return false;
+		}
+		// If we get a measurement, we take it as soon as we can
+		_execute_queued_ops(target, OP::Ry);
+		// If we measure 0, then this resets the occupied qubit register
+		if (_quantum_state->M(target)){
+			_set_qubit_to_nonzero(target);
+		} else {
+			_set_qubit_to_zero(target);
+		}
+		return _occupied_qubits[target];
+
+	}
+
+	void Reset(logical_qubit_id target) {
+		if (!_occupied_qubits[target]){ return; }
+		// If we get a measurement, we take it as soon as we can
+			_execute_queued_ops(target, OP::Ry);
+		_quantum_state->Reset(target);
+		_set_qubit_to_zero(target);
+	}
+
+	void Assert(std::vector<Gates::Basis> axes, std::vector<logical_qubit_id> const& qubits, bool result) {
```

Comment:
```suggestion
	void Assert(std::vector<Gates::Basis> const& axes, std::vector<logical_qubit_id> const& qubits, bool result) {
```

Reply:
This won't work. This array gets updated in the function.